packages/utils/src/misc.ts

```diff
@@ -86,23 +89,47 @@ export function getRelayClientMetadata(protocol: string, version: number): Relay
 
   const metadata: RelayerClientMetadata = { protocol, version, env };
   if (env === "browser") {
-    metadata.host = getLocation()?.host || "";
+    metadata.host = getLocation()?.host || "unknown";
   }
   return metadata;
 }
 
 // -- rpcUrl ----------------------------------------------//
 
+export function getJavascriptOS() {
+  const info = detect();
+  if (info === null) return "unknown";
+  const os = info.os ? info.os.replace(" ", "").toLowerCase() : "unknown";
+  if (info.type === "browser") {
+    return [os, info.name, info.version].join("-");
+  }
+  return [os, info.version].join("-");
+}
+
+export function getJavascriptID() {
+  const env = getEnvironment();
+  return env === ENV_MAP.browser ? [env, getLocation()?.host || "unknown"].join(":") : env;
+}
+
+export function formatUA(protocol: string, version: number, sdkVersion: string) {
+  const os = getJavascriptOS();
+  const id = getJavascriptID();
+  return [[protocol, version].join("-"), [SDK_TYPE, sdkVersion].join("-"), os, id].join("/");
+}
+console;
+
 export function formatRelayRpcUrl({
   protocol,
   version,
   relayUrl,
+  sdkVersion,
   auth,
   projectId,
 }: RelayerTypes.RpcUrlParams) {
   const splitUrl = relayUrl.split("?");
-  const metadata = { ...getRelayClientMetadata(protocol, version), auth };
-  const params = projectId ? { ...metadata, projectId } : metadata;
+  const ua = formatUA(protocol, version, sdkVersion);
+  const params = { auth, ua, projectId };
```

Comment:
Minor: Think we can save ourselves the inline `if` with object mutation if we keep the ternary style:

```suggestion
  const params = projectId ? { auth, ua, projectId } : { auth, ua };
```

Reply:
honestly we should have a util that sanitizes "undefined" values instead because what I don't like about the ternary approach is that you have to duplicate the object for the other values

However this delete might be unnecessary too... if the "query-string" package already sanitizes undefined values

Reply:
> what I don't like about the ternary approach is that you have to duplicate the object for the other values

Yeah I get that too, neither approach is ideal in the end. Let's just leave it as is, mostly stylistic anyways 👍